<center> 
    <h1> Toronto Crime </h1>
<h3> Sprint 3</h3>
<h3> More Basic Modelling</h3>
    </center>

##### Author: Arsalan Arif Radhu
##### Date: 21 November 2023

### Table of Contents:
- [Introduction](#intro) <br />
- [KNN](#knn)<br />

### Introduction <a id="intro" /> 

In [ ]:
# Import necessary libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from xgboost import XGBClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report
from matplotlib import pyplot as plt

In [ ]:
# Importing the data
df = pd.read_csv('../Data/FinalDataAdvAn.csv',index_col=0)
df.reset_index(inplace=True)

In [ ]:
df.shape

(414327, 39)

Dropping the `UCR_CODE` and `UCR_EXT` as I just realized that these indicate which crime type has occurred. Hence, I will be dropping them.

In [ ]:
df = df.drop('UCR_CODE', axis=1)
df = df.drop('UCR_EXT', axis=1)

In [ ]:
df.columns

Index(['OBJECTID', 'CRIME_TYPE', 'REPORT_YEAR', 'REPORT_MONTH', 'REPORT_DAY',
       'OCC_YEAR', 'OCC_MONTH', 'OCC_DAY', 'LONG_WGS84', 'LAT_WGS84', 'DEATH',
       'INJURIES', 'D11', 'D12', 'D13', 'D14', 'D22', 'D23', 'D31', 'D32',
       'D33', 'D41', 'D42', 'D43', 'D51', 'D52', 'D53', 'D55', 'NSA',
       'Apartment', 'Commercial', 'Educational', 'House', 'NotApplicable',
       'Other', 'Outside', 'Transit'],
      dtype='object')

In [ ]:
df.sample(10)

OBJECTID  CRIME_TYPE  REPORT_YEAR  REPORT_MONTH  REPORT_DAY  OCC_YEAR  \
117299    128510           0         2020            10           6      2020   
278015     47732           5         2019             6          11      2019   
33045      36295           0         2016             1          30      2016   
150636    164710           0         2022             8          14      2022   
128906    141092           0         2021             6          24      2021   
332834     24574           3         2020             8          16      2020   
56737      62162           0         2017             6          12      2017   
9239       10174           0         2014             8          15      2014   
378586     26695           1         2017            11          20      2017   
288881     58614           5         2020             6           8      2020   

        OCC_MONTH  OCC_DAY  LONG_WGS84  LAT_WGS84  ...  D55  NSA  Apartment  \
117299         10        6  -79.199469  43.765441  ...    0    0          1   
278015          6        9  -79.344687  43.687372  ...    1    0          0   
33045           1       30  -79.253430  43.737863  ...    0    0          0   
150636          8       14  -79.446110  43.640305  ...    0    0          0   
128906          6       24  -79.423550  43.639717  ...    0    0          0   
332834          8       16  -79.358869  43.676196  ...    1    0          0   
56737           6       12  -79.438787  43.756785  ...    0    0          1   
9239            7       31  -79.281937  43.798033  ...    0    0          0   
378586         11       20  -79.271107  43.767919  ...    0    0          0   
288881          6        7  -79.411614  43.685692  ...    0    0          0   

        Commercial  Educational  House  NotApplicable  Other  Outside  Transit  
117299           0            0      0              0      0        0        0  
278015           0            0      0              0      0        1        0  
33045            0            1      0              0      0        0        0  
150636           0            0      0              0      0        1        0  
128906           1            0      0              0      0        0        0  
332834           1            0      0              0      0        0        0  
56737            0            0      0              0      0        0        0  
9239             0            0      1              0      0        0        0  
378586           1            0      0              0      0        0        0  
288881           0            0      0              0      0        1        0  

[10 rows x 37 columns]

## KNN <a id="knn" /> 


I am trying out k-Nearest Neighbors (KNN) algorithm due to its simplicity and versatility, making it an ideal choice for my current machine learning task. As a non-parametric method, KNN allows me to work with data without the need to assume a specific distribution, offering flexibility in handling diverse datasets. The absence of a formal training phase is advantageous for my dynamic dataset, allowing the model to adapt swiftly to changes. Additionally, KNN's suitability for small datasets, interpretability, and robustness to outliers align with the characteristics of my specific data, making it a practical and effective choice for my analysis.

In [ ]:
# Extract features and labels
X = df.drop('CRIME_TYPE', axis=1)  # Replace 'label_column' with the actual label column name
y = df['CRIME_TYPE']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=30)

# Initialize the KNN classifier
knn_classifier = KNeighborsClassifier(n_neighbors=5)  # You can adjust the number of neighbors (k) as needed

# Train the model
knn_classifier.fit(X_train, y_train)

# Make predictions on the test set
y_pred = knn_classifier.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
classification_rep = classification_report(y_test, y_pred)

# Display the results
print(f"Accuracy: {accuracy}")
print("Classification Report:")
print(classification_rep)

Accuracy: 0.9540704269543601
Classification Report:
              precision    recall  f1-score   support

           0       0.98      0.98      0.98     31647
           1       0.94      0.94      0.94     12347
           2       0.77      0.70      0.73       234
           3       0.89      0.93      0.91      5246
           4       0.94      0.43      0.59      1204
           5       0.95      0.96      0.95     32188

    accuracy                           0.95     82866
   macro avg       0.91      0.82      0.85     82866
weighted avg       0.95      0.95      0.95     82866




The model achieved an overall accuracy of <b>95.4%</b>, showcasing strong performance in classifying instances across multiple categories. Notably, it demonstrated high precision and recall for the majority of classes, with particularly robust results for class 0, achieving a precision and recall of 98%. 
But it is too accurate leading me to believe that it might be overfitting or there might be an issue with my data set.